app/code/Magento/InventoryCatalog/Model/GetProductIdsBySkus.php

```diff
@@ -40,21 +40,32 @@ public function __construct(
      */
     public function execute(array $skus): array
     {
-        $cacheKey = hash('md5', implode(',', $skus));
+        $result = [];
 
-        if (!isset($this->productIdsBySkus[$cacheKey])) {
-            $idsBySkus = $this->productResource->getProductsIdsBySkus($skus);
-            $notFoundedSkus = array_diff($skus, array_keys($idsBySkus));
-
-            if (!empty($notFoundedSkus)) {
-                throw new NoSuchEntityException(
-                    __('Following products with requested skus were not found: %1', implode($notFoundedSkus, ', '))
-                );
+        foreach ($skus as $index => $sku) {
+            if (!isset($this->idsBySku[$sku])) {
+                continue;
             }
 
-            $this->productIdsBySkus[$cacheKey] = $idsBySkus;
+            unset($skus[$index]);
+            $result[$sku] = $this->productIdsBySkus[$sku];
+        }
+
+        if (empty($skus)) {
+            return $result;
         }
 
-        return $this->productIdsBySkus[$cacheKey];
+        $result = array_replace($result, $this->productResource->getProductsIdsBySkus($skus));
+        $notFoundSkus = array_diff($skus, array_keys($result));
+
+        if (!empty($notFoundSkus)) {
+            throw new NoSuchEntityException(
+                __('Following products with requested skus were not found: %1', implode($notFoundSkus, ', '))
+            );
+        }
+
+        $this->productIdsBySkus = array_replace($this->productIdsBySkus, $result);
+
+        return $result;
```

Comment:
@maghamed 

> This is the most interesting question. I was thinking a bit about it while reviewing PR #1960
My point here is that taking into account that our cache is valid just in a scope of single HTTP request, it does not make sense to introduce more sophisticated caching layer which is able to store data on the very low granularity level (per one product SKU -> ID). Because we don't have such business scenarious where in the scope of one HTTP request processing we need to resolve SKUs -> IDs on the bulk level, and then apply resolving item by item.
If we had a caching layer that supposed to be re-used by different HTTP requests - I am totally agree that in this particular case we have to deal with the most granular level possible.
But introducing this verbosity we can get some extra complexity now, especially working with Bulk resolving.

I thought; since it's a single HTTP request there is a good chance that the same sku might be used in different ways. It kind of depends on how the `IsProductSalableConditionChain` is called and whether the result is 'cached'. But is seems to be called from multiple plugins so assume there is a good change that it might be called with a collection of sku's and individual sku's. I would be more concerned about bulk actions since there might be more processing power used to loop over all the items compared to doing a query.